src/mscorlib/shared/System/IO/Path.cs

```diff
@@ -76,19 +76,27 @@ public static string ChangeExtension(string path, string extension)
         // "\\server\share").
         public static string GetDirectoryName(string path)
         {
-            if (path != null)
+            if (path == null)
             {
-                PathInternal.CheckInvalidPathChars(path);
-                path = PathInternal.NormalizeDirectorySeparators(path);
-                int root = PathInternal.GetRootLength(path);
+                return null;
+            }
 
-                int i = path.Length;
-                if (i > root)
-                {
-                    while (i > root && !PathInternal.IsDirectorySeparator(path[--i])) ;
-                    return path.Substring(0, i);
-                }
+            if (string.IsNullOrWhiteSpace(path))
+            {
+                throw new ArgumentException(SR.Arg_PathIllegal, nameof(path));
+            }
```

Comment:
It'd be better to make this something like:
```C#
if (string.IsNullOrWhitespace(path))
{
    if (path == null) return null;
    throw new ArgumentException(SR.Arg_PathIllegal, nameof(path));
}
```
That way we only do the null check on path once in the common path rather than twice.